README.md
CSV Reader
==========

The [CsvReader](https://www.nuget.org/packages/LumenWorksCsvReader/) library is an extended version of Sébastien Lorion's [fast CSV Reader](http://www.codeproject.com/Articles/9258/A-Fast-CSV-Reader) project 
and provides fast parsing and reading of CSV files

[![NuGet](https://img.shields.io/nuget/v/LumenWorksCsvReader.svg)](https://www.nuget.org/packages/LumenWorksCsvReader/)
[![Build status](https://ci.appveyor.com/api/projects/status/ouvglmaox83bpyti/branch/master?svg=true)](https://ci.appveyor.com/project/PaulHatcher/csvreader/branch/master)

To this end it is a straight drop-in replacement for the existing NuGet package [LumenWorks.Framework.IO](https://www.nuget.org/packages/LumenWorks.Framework.IO/), but with additional
capabilities; the other rationale for the project is that the code is not available elsewhere in a public source repository, making it difficult to extend/contribute to.

Welcome to contributions from anyone.

You can see the version history [here](RELEASE_NOTES.md).

## Build the project
* Windows: Run *build.cmd*

The tooling should be automatically installed by paket/Fake. The default build will compile and test the project, and also produce a nuget package.

The library supports for .NET 2.0, 3.5, 4.5 and 4.6.1 and .netstandard 1.3 and 2.0, the netstandard1.3 version does not contain CachedCsvReader since the necessary interfaces are not available.

## Library License

The library is available under the [MIT License](http://en.wikipedia.org/wiki/MIT_License), for more information see the [License file][1] in the GitHub repository.

 [1]: https://github.com/phatcher/CsvReader/blob/master/License.md

## Getting Started
A good starting point is to look at Sébastien's [article](http://www.codeproject.com/Articles/9258/A-Fast-CSV-Reader) on Code Project.

A basic use of the reader something like this...
```csharp
using System.IO;
using LumenWorks.Framework.IO.Csv;

void ReadCsv()
{
    // open the file "data.csv" which is a CSV file with headers
    using (var csv = new CachedCsvReader(new StreamReader("data.csv"), true))
    {
        // Field headers will automatically be used as column names
        myDataGrid.DataSource = csv;
    }
}
```
Having said that, there are some extensions built into this version of the library that it is worth mentioning.

## Additional Features

### Columns
One addition is the addition of a Column list which holds the names and types of the data in the CSV file. If there are no headers present, we default the column names to Column1, Column2 etc; this can be overridden by setting the DefaultColumnHeader property e.g.
```csharp
using System.IO;
using LumenWorks.Framework.IO.Csv;

void ReadCsv()
{
    // open the file "data.csv" which is a CSV file without headers
    using (var csv = new CachedCsvReader(new StreamReader("data.csv"), false))
    {
        csv.DefaultColumnHeader = "Fred"

        // Field headers will now be Fred1, Fred2, etc
        myDataGrid.DataSource = csv;
    }
}
```

You can specify the columns yourself if there are none, and also specify the expected type; this is especially important when using against SqlBulkCopy which we will come back to later.
```csharp
using System.IO;
using LumenWorks.Framework.IO.Csv;

void ReadCsv()
{
    // open the file "data.csv" which is a CSV file without headers
    using (var csv = new CachedCsvReader(new StreamReader("data.csv"), false))
    {
        csv.Columns.Add(new Column { Name = "PriceDate", Type = typeof(DateTime) });
        csv.Columns.Add(new Column { Name = "OpenPrice", Type = typeof(decimal) });
        csv.Columns.Add(new Column { Name = "HighPrice", Type = typeof(decimal) });
        csv.Columns.Add(new Column { Name = "LowPrice", Type = typeof(decimal) });
        csv.Columns.Add(new Column { Name = "ClosePrice", Type = typeof(decimal) });
        csv.Columns.Add(new Column { Name = "Volume", Type = typeof(int) });

        // Field headers will now be picked from the Columns collection
        myDataGrid.DataSource = csv;
    }
}
```

### SQL Bulk Copy
One use of CSV Reader is to have a nice .NET way of using SQL Bulk Copy (SBC) rather than bcp for bulk loading of data into SQL Server.

A couple of issues arise when using SBC
    1. SBC wants the data presented as the correct type rather than as string
    2. You need to map between the table destination columns and the CSV if the order does not match *exactly*

Below is a example using the Columns collection to set up the correct metadata for SBC
```csharp
public void Import(string fileName, string connectionString)
{
    using (var reader = new CsvReader(new StreamReader(fileName), false))
    {
        reader.Columns = new List<LumenWorks.Framework.IO.Csv.Column>
        {
            new LumenWorks.Framework.IO.Csv.Column { Name = "PriceDate", Type = typeof(DateTime) },
            new LumenWorks.Framework.IO.Csv.Column { Name = "OpenPrice", Type = typeof(decimal) },
            new LumenWorks.Framework.IO.Csv.Column { Name = "HighPrice", Type = typeof(decimal) },
            new LumenWorks.Framework.IO.Csv.Column { Name = "LowPrice", Type = typeof(decimal) },
            new LumenWorks.Framework.IO.Csv.Column { Name = "ClosePrice", Type = typeof(decimal) },
            new LumenWorks.Framework.IO.Csv.Column { Name = "Volume", Type = typeof(int) },
        };

        // Now use SQL Bulk Copy to move the data
        using (var sbc = new SqlBulkCopy(connectionString))
        {
            sbc.DestinationTableName = "dbo.DailyPrice";
            sbc.BatchSize = 1000;

            sbc.AddColumnMapping("PriceDate", "PriceDate");
            sbc.AddColumnMapping("OpenPrice", "OpenPrice");
            sbc.AddColumnMapping("HighPrice", "HighPrice");
            sbc.AddColumnMapping("LowPrice", "LowPrice");
            sbc.AddColumnMapping("ClosePrice", "ClosePrice");
            sbc.AddColumnMapping("Volume", "Volume");

            sbc.WriteToServer(reader);
        }
    }
}
```
The method AddColumnMapping is an extension I wrote to simplify adding mappings to SBC
```csharp
public static class SqlBulkCopyExtensions
{
    public static SqlBulkCopyColumnMapping AddColumnMapping(this SqlBulkCopy sbc, int sourceColumnOrdinal, int targetColumnOrdinal)
    {
        var map = new SqlBulkCopyColumnMapping(sourceColumnOrdinal, targetColumnOrdinal);
        sbc.ColumnMappings.Add(map);

        return map;
    }

    public static SqlBulkCopyColumnMapping AddColumnMapping(this SqlBulkCopy sbc, string sourceColumn, string targetColumn)
    {
        var map = new SqlBulkCopyColumnMapping(sourceColumn, targetColumn);
        sbc.ColumnMappings.Add(map);

        return map;
    }
}
```

One other issue recently arose where we wanted to use SBC but some of the data was not in the file itself, but metadata that needed to be included on every row. The solution was to amend the CSV reader and Columns collection to allow default values to be provided that are not in the data.

The additional columns should be added at the end of the Columns collection to avoid interfering with the parsing, see the amended example below...
```csharp
public void Import(string fileName, string connectionString)
{
    using (var reader = new CsvReader(new StreamReader(fileName), false))
    {
        reader.Columns = new List<LumenWorks.Framework.IO.Csv.Column>
        {
            ...
            new LumenWorks.Framework.IO.Csv.Column { Name = "Volume", Type = typeof(int) },
            // NB Fake column so bulk import works
            new LumenWorks.Framework.IO.Csv.Column { Name = "Ticker", Type = typeof(string) },
        };

        // Fix up the column defaults with the values we need
        reader.UseColumnDefaults = true;
        reader.Columns[reader.GetFieldIndex("Ticker")] = Path.GetFileNameWithoutExtension(fileName);

        // Now use SQL Bulk Copy to move the data
        using (var sbc = new SqlBulkCopy(connectionString))
        {
            ...
            sbc.AddColumnMapping("Ticker", "Ticker");

            sbc.WriteToServer(reader);
        }
    }
}
```
To give an idea of performance, this took a naive sample app using an ORM from 2m 27s to 1.37s using SBC and the full import took just over 11m to import 9.8m records.

### Null Byte Removal StreamReader
Use NullRemovalStreamReader when CSV files contain large number of null bytes and you do not control how to generate CSV files.

If you ever experienced "System.OutOfMemoryException" or long processing time, you will most likely get a huge performance gain with NullRemovalStreamReader.
```csharp
public void Process(string path, bool addMark)
{
    using (StreamReader stream = new StreamReader(path))
    using (CsvReader csv = new CsvReader(stream.BaseStream, false, stream.CurrentEncoding, addMark))
    // or using (CsvReader csv = new CsvReader(File.OpenRead(path), false, Encoding.UTF8, addMark))
    {
        while (csv.ReadNextRecord())
        {
            string data = csv[i];
            // do stuff
        }
    }
}
```

When addMark is true, consecutive null bytes will be replaced by [removed x null bytes] to indicate the removal, you can see this from the benchmark output below.

Performance differences shown when tested with 20 million null bytes (20MB in storage) :
```csharp
CsvReader -     without using NullRemovalStreamReader : 536968 ticks, 0.2120 sec., 94.3518 MB/sec.

CsvReader - with NullRemovalStreamReader without mark : 191137 ticks, 0.0755 sec., 265.0660 MB/sec.
AddMark =(False) LastCell =(cell63 followed by 20971520 null bytes)

CsvReader - with NullRemovalStreamReader with    mark : 168819 ticks, 0.0666 sec., 300.1079 MB/sec.
AddMark =(True) LastCell =(cell63 followed by 20971520 null bytes[removed 20971520 null bytes])
```

Adjust number of null bytes in benchmark to see how much memory/time you will be able to save:
```csharp
X:\Path\CsvReader\build\Debug\CsvReaderBenchmarks\net461>CsvReaderBenchmarks.exe NullRemoval
```


## Performance
One of the main reasons for using this library is its excellent performance on reading/parsing raw data, here's a recent run of the benchmark (which is in the source)
```csharp
Test pass #1 - All fields

CsvReader - No cache      : 3134597 ticks, 1.2374 sec., 35.5582 MB/sec.
CachedCsvReader - Run 1   : 7452030 ticks, 2.9418 sec., 14.9571 MB/sec.
CachedCsvReader - Run 2   : 4525 ticks, 0.0018 sec., 24632.1821 MB/sec.
TextFieldParser           : 31568009 ticks, 12.4617 sec., 3.5308 MB/sec.
Regex                     : 11273590 ticks, 4.4503 sec., 9.8869 MB/sec.

Test pass #1 - Field #72 (middle)

CsvReader - No cache      : 2358656 ticks, 0.9311 sec., 47.2560 MB/sec.
CachedCsvReader - Run 1   : 7119186 ticks, 2.8104 sec., 15.6564 MB/sec.
CachedCsvReader - Run 2   : 325 ticks, 0.0001 sec., 342955.7662 MB/sec.
TextFieldParser           : 31171440 ticks, 12.3052 sec., 3.5757 MB/sec.
Regex                     : 5793093 ticks, 2.2869 sec., 19.2403 MB/sec.


Test pass #2 - All fields

CsvReader - No cache      : 2941954 ticks, 1.1614 sec., 37.8866 MB/sec.
CachedCsvReader - Run 1   : 7204077 ticks, 2.8439 sec., 15.4719 MB/sec.
CachedCsvReader - Run 2   : 314 ticks, 0.0001 sec., 354970.1401 MB/sec.
TextFieldParser           : 31213609 ticks, 12.3218 sec., 3.5709 MB/sec.
Regex                     : 11095897 ticks, 4.3802 sec., 10.0452 MB/sec.

Test pass #2 - Field #72 (middle)

CsvReader - No cache      : 2186909 ticks, 0.8633 sec., 50.9672 MB/sec.
CachedCsvReader - Run 1   : 7131654 ticks, 2.8153 sec., 15.6290 MB/sec.
CachedCsvReader - Run 2   : 296 ticks, 0.0001 sec., 376556.1622 MB/sec.
TextFieldParser           : 31381026 ticks, 12.3879 sec., 3.5518 MB/sec.
Regex                     : 5151353 ticks, 2.0335 sec., 21.6372 MB/sec.


Test pass #3 - All fields

CsvReader - No cache      : 2693834 ticks, 1.0634 sec., 41.3762 MB/sec.
CachedCsvReader - Run 1   : 7105358 ticks, 2.8049 sec., 15.6868 MB/sec.
CachedCsvReader - Run 2   : 326 ticks, 0.0001 sec., 341903.7546 MB/sec.
TextFieldParser           : 31323784 ticks, 12.3653 sec., 3.5583 MB/sec.
Regex                     : 11303752 ticks, 4.4622 sec., 9.8605 MB/sec.

Test pass #3 - Field #72 (middle)

CsvReader - No cache      : 2177773 ticks, 0.8597 sec., 51.1810 MB/sec.
CachedCsvReader - Run 1   : 7326816 ticks, 2.8923 sec., 15.2127 MB/sec.
CachedCsvReader - Run 2   : 328 ticks, 0.0001 sec., 339818.9756 MB/sec.
TextFieldParser           : 31168390 ticks, 12.3040 sec., 3.5761 MB/sec.
Regex                     : 5134853 ticks, 2.0270 sec., 21.7067 MB/sec.


Done
```
This was run on a high-spec machine (Xeon E5-2640, 32Gb RAM and M.2 1Tb SSD; you have to have some toys!) so the overall thoughput would be good, but CsvReader performs at 10x the speed the TextFieldParser and 5x faster than Regex


RELEASE_NOTES.md
#### 4.0.1 (2018-10-21)
* Introduce SourceLink support

#### 4.0.0 (2018-03-24)
* Support .NET Framework 2.0+ and .NET Standard 1.3+
* Use DateTime.TryParseExact if needed (@milcondoin)
* Override column capability (@spintronic)
* Fix handling of duplicate headers (@jonreis)
* Reduced exception overhead (@spintronic)
* Added NullBytesRemoval functionality to reduce memory usage and runtime (@andrewpsy)
* Use StringBuilder instead of string concatenation to speed up reading of huge cell (@andrewpsy)
* Fixed: Exception in GetFieldType() when hasHeaders is false (@molopony)
* Fixed: Exception when reading data and hasHeaders is false (@molopony)
* Fixed: ArgumentOutOfRangeException when hasHeaders is false (@molopony)

#### 3.9.1 (2016-04-09)
* Added max field length for quoted fields (@criteo)
* Strengthen Debug.Assert condition in GetSchemaWithoutHeaders (@joshgo)
* Added test for quoted fields not throwing an exception (@joshgo)
* Added HasHeader function to allow checking if a header existing without throwing exception (@kiran94)

#### 3.9 (2015-06-20)
* Completely replaced use of _fieldHeaders internally with Columns
* Fixed IDataReader.GetSchemaTable to return Column.Type rather than always saying string
* Introduced UseColumnDefault to allow Column.Default value to be returned when a column doesn't exist in the source data.
* Changing handling of extra fields to respect DefaultParseErrorAction (thanks to @gap777)
* Added support for null values and some tests (thanks to @fretje)

#### 3.8.3 (2015-03-21)
* Can now handle binary and varbinary columsn (thanks to @fretje)

#### 3.8.2 (2014-02-27)
* Adding Column metadata so we can use with SqlBulkCopy.

#### 3.8.1 (2011-11-10)
* Fixed bug with missing field handling.
* Converted solution to VS 2010 (still targets .NET 2.0).

#### 3.8 (2011-07-05)
* Empty header names in CSV files are now replaced by a default name that can be customized via the new DefaultHeaderName property (by default, it is "Column" + column index).

#### 3.7.2 (2011-05-17)
* Fixed a bug when handling missing fields.
* Strongly named the main assembly.

#### 3.7.1 (2010-11-03)
* Fixed a bug when handling whitespaces at the end of a file.

#### 3.7 (2010-03-30)
* Breaking: Added more field value trimming options.

#### 3.6.2 (2008-10-09)
* Fixed a bug when calling MoveTo in a particular action sequence;
* Fixed a bug when extra fields are present in a multiline record;
* Fixed a bug when there is a parse error while initializing.

#### 3.6.1 (2008-07-16)
* Fixed a bug with RecordEnumerator caused by reusing the same array over each iteration.

#### 3.6 (2008-07-09)
* Added a web demo project;
* Fixed a bug when loading CachedCsvReader into a DataTable and the CSV has no header.

#### 3.5 (2007-11-28)
* Fixed a bug when initializing CachedCsvReader without having read a record first.

#### 3.4 (2007-10-23)
* Fixed a bug with the IDataRecord implementation where GetValue/GetValues should return DBNull.Value when the field value is empty or null;
* Fixed a bug where no exception is raised if a delimiter is not present after a non final quoted field;
* Fixed a bug when trimming unquoted fields and whitespaces span over two buffers.

#### 3.3 (2007-01-14)
* Added the option to turn off skipping empty lines via the property SkipEmptyLines (on by default);
* Fixed a bug with the handling of a delimiter at the end of a record preceded by a quoted field.

#### 3.2 (2006-12-11)
* Slightly modified the way missing fields are handled;
* Fixed a bug where the call to CsvReader.ReadNextRecord() would return false for a CSV file containing only one line ending with a new line character and no header.

#### 3.1.2 (2006-08-06)
* Updated dispose pattern;
* Fixed a bug when SupportsMultiline is false;
* Fixed a bug where the IDataReader schema column "DataType" returned DbType.String instead of typeof(string).

#### 3.1.1 (2006-07-25)
* Added a SupportsMultiline property to help boost performance when multi-line support is not needed;
* Added two new constructors to support common scenarios;
* Added support for when the base stream returns a length of 0;
* Fixed a bug when the FieldCount property is accessed before having read any record;
* Fixed a bug when the delimiter is a whitespace;
* Fixed a bug in ReadNextRecord(...) by eliminating its recursive behavior when initializing headers;
* Fixed a bug when EOF is reached when reading the first record;
* Fixed a bug where no exception would be thrown if the reader has reached EOF and a field is missing.

#### 3.0 (2006-05-15)
* Introduced equal support for .NET 1.1 and .NET 2.0;
* Added extensive support for malformed CSV files;
* Added complete support for data-binding;
* Made available the current raw data;
* Field headers are now accessed via an array (breaking change);
* Made field headers case insensitive (thanks to Marco Dissel for the suggestion);
* Relaxed restrictions when the reader has been disposed;
* CsvReader supports 2^63 records;
* Added more test coverage;
* Upgraded to .NET 2.0 release version;
* Fixed an issue when accessing certain properties without having read any data (notably FieldHeaders).

#### 2.0 (2005-08-10)
* Ported code to .NET 2.0 (July 2005 CTP);
* Thoroughly debugged via extensive unit testing (special thanks to shriop);
* Improved speed (now 15 times faster than OLEDB);
* Consumes half the memory than version 1.0;
* Can specify a custom buffer size;
* Full Unicode support;
* Auto-detects line ending, be it \r, \n, or \r\n;
* Better exception handling;
* Supports the "field1\rfield2\rfield3\n" pattern (used by Unix);
* Parsing code completely refactored, resulting in much cleaner code.

#### 1.1 (2005-01-15)
* 1.1: Added support for multi-line fields.

#### 1.0 (2005-01-09)
* 1.0: First release.

code/CsvReaderBenchmarks/Program.cs
using System;
using System.IO;

namespace CsvReaderBenchmarks
{
    internal class Program
    {
        // StopWatch seems to not be accurate enough to be used here (divisions by zero occur when calculating MB/s).

        [System.Runtime.InteropServices.DllImport("Kernel32.dll")]
        private static extern bool QueryPerformanceCounter(out long lpPerformanceCount);

        [System.Runtime.InteropServices.DllImport("Kernel32.dll")]
        private static extern bool QueryPerformanceFrequency(out long lpFrequency);

        [STAThread()]
        private static void Main(string[] args)
        {
            const string TestFile2 = @"test2.csv";
            const string TestFile3 = @"test3.csv";

#if !NETCOREAPP1_0
            AppDomain.CurrentDomain.UnhandledException += CurrentDomain_UnhandledException;
#endif

            if (args.Length > 0)
            {
                if (args.Length == 1)
                {
                    switch (args[0].ToUpperInvariant())
                    {
                        case "CSVREADER":
                            CsvReaderBenchmark.Run(TestFile3);
                            return;

                        case "NULLREMOVAL":
                            PerformanceTestWithNullRemovalStreamReader();
                            return;

                        case "STRINGBUILDER":
                            PerformanceTestWithStringBuilder();
                            return;

#if !NETCOREAPP1_0 && !NETCOREAPP2_0
                        case "OLEDB":
                            OleDbBenchmark.Run(TestFile3);
                            return;
#endif
                        case "REGEX":
                            RegexBenchmark.Run(TestFile3);
                            return;

                        default:
                            Console.WriteLine(@"Possible values: CsvReader, NullRemoval, StringBuilder, OleDb, Regex");
                            break;
                    }
                }

                return;
            }

            const int field = 72;
            var fileSize = new FileInfo(TestFile2).Length / 1024 / 1024;

            for (var i = 1; i < 4; i++)
            {
                Console.WriteLine("Test pass #{0} - All fields\n", i);

                DoTest("CsvReader - No cache", fileSize, CsvReaderBenchmark.Run, TestFile2);
#if !NETCOREAPP1_0
                object csv = DoTest("CachedCsvReader - Run 1", fileSize, CachedCsvReaderBenchmark.Run1, TestFile2);
                DoTest("CachedCsvReader - Run 2", fileSize, CachedCsvReaderBenchmark.Run2, csv);
#endif
#if !NETCOREAPP1_0 && !NETCOREAPP2_0
                DoTest("TextFieldParser", fileSize, TextFieldParserBenchmark.Run, TestFile2);
#endif
                DoTest("Regex", fileSize, RegexBenchmark.Run, TestFile2);

                // seems to not be working on Windows 7 with Office 2007 (and I'm not bothering to try to make it run on my machine)
                //DoTest("OleDb", fileSize, OleDbBenchmark.Run, TestFile2);

                Console.WriteLine();

                Console.WriteLine("Test pass #{0} - Field #{1} (middle)\n", i, field);

                DoTest("CsvReader - No cache", fileSize, CsvReaderBenchmark.Run, TestFile2, field);
#if !NETCOREAPP1_0
                csv = DoTest("CachedCsvReader - Run 1", fileSize, CachedCsvReaderBenchmark.Run1, TestFile2, field);
                DoTest("CachedCsvReader - Run 2", fileSize, CachedCsvReaderBenchmark.Run2, csv, field);
#endif
#if !NETCOREAPP1_0 && !NETCOREAPP2_0
                DoTest("TextFieldParser", fileSize, TextFieldParserBenchmark.Run, TestFile2, field);
#endif
                DoTest("Regex", fileSize, RegexBenchmark.Run, TestFile2, field);

                // seems to not be working on Windows 7 with Office 2007 (and I'm not bothering to try to make it run on my machine)
#if !NETCOREAPP1_0 && !NETCOREAPP2_0
                //DoTest("OleDb", fileSize, OleDbBenchmark.Run, TestFile2, Field);
#endif

                Console.WriteLine();
                Console.WriteLine();
            }

            Console.WriteLine("Done");
            Console.ReadLine();
        }

        private delegate object TestCallback(object[] args);

        private static object DoTest(string name, long fileSize, TestCallback testCallback, params object[] args)
        {
            QueryPerformanceFrequency(out long frequency);

            GC.Collect();
            GC.WaitForPendingFinalizers();
            GC.Collect();

            QueryPerformanceCounter(out long start);
            var value = testCallback(args);
            QueryPerformanceCounter(out long end);
            GetStats(start, end, frequency, fileSize, out long clocks, out double time, out double rate);

            Console.WriteLine("{0} : {1} ticks, {2:f4} sec., {3:f4} MB/sec.", name.PadRight(25), clocks, time, rate);

            return value;
        }

        private static void GetStats(long start, long end, long frequency, long fileSize, out long clocks, out double time, out double rate)
        {
            clocks = end - start;
            time = (double) clocks / frequency;
            rate = fileSize / time;
        }

#if !NETCOREAPP1_0
        private static void CurrentDomain_UnhandledException(object sender, UnhandledExceptionEventArgs e)
        {
            if (e.ExceptionObject != null)
            {
                Console.WriteLine("Unhandled exception :\n\n'{0}'.", e.ExceptionObject);
            }
            else
            {
                Console.WriteLine("Unhandled exception occured.");
            }
            Console.ReadLine();
        }
#endif

        private static void PerformanceTestWithNullRemovalStreamReader()
        {
            var path = string.Empty;
            try
            {
                path = GenerateCsvFile();
                var fileSize = new FileInfo(path).Length / 1024 / 1024;
                DoTest("CsvReader -     without using NullRemovalStreamReader", fileSize, CsvReaderBenchmark.Run, path);
                Console.WriteLine();
                var result = DoTest("CsvReader - with NullRemovalStreamReader without mark", fileSize, CsvReaderBenchmark.Run, path, -1, false);
                Console.WriteLine(result + Environment.NewLine);
                result = DoTest("CsvReader - with NullRemovalStreamReader with    mark", fileSize, CsvReaderBenchmark.Run, path, -1, true);
                Console.WriteLine(result);
            }
            finally
            {
                if (!string.IsNullOrEmpty(path) && File.Exists(path))
                {
                    File.Delete(path);
                }
            }
        }

        private static void PerformanceTestWithStringBuilder()
        {
            var path = string.Empty;
            try
            {
                path = GenerateCsvFile(false, 5);
                var fileSize = new FileInfo(path).Length / 1024 / 1024;
                DoTest("CsvReader - performance test with large cell", fileSize, CsvReaderBenchmark.Run, path);
            }
            finally
            {
                if (!string.IsNullOrEmpty(path) && File.Exists(path))
                {
                    File.Delete(path);
                }
            }
        }

        private static string GenerateCsvFile(bool testNull = true, int x = 20)
        {
            // generate x million bytes; file size will be a little over x MB
            long count = x * 1024 * 1024;
            var path = Path.GetTempFileName();

            using (var sw = File.AppendText(path))
            {
                for (var i = 1; i <= 5; i++)
                {
                    sw.WriteLine("cell{0}1,cell{0}2,cell{0}3", i);
                }
                if (testNull)
                {
                    sw.Write("cell61,cell62,cell63 followed by " + count + " null bytes");
                    sw.Write(new char[count]);
                }
                else
                {
                    sw.Write("cell61,cell62,cell63 followed by " + count + " 'a' characters=>");
                    var data = new char [count];
                    for (var i = 0; i < count; i++)
                    {
                        data[i] = 'a';
                    }
                    sw.Write(data);
                }
            }
            return path;
        }
    }
}
